File: generate_prompts.py
import os
import json
from pathlib import Path

def generate_prompts():
    project_dir = Path(r"f:\Projects\koala\works\穿书后我攻略了奸臣首辅")
    shots_dir = project_dir / "scripts" / "episode_001" / "shots"
    output_dir = project_dir / "output" / "episode_001"
    
    with open(project_dir / "project.json", "r", encoding="utf-8") as f:
        project_data = json.load(f)
        global_style = project_data.get("style", {}).get("globalPromptStyle", "古风动漫风格")
        
    for shot_idx in range(1, 5):
        shot_id = f"shot_{shot_idx:03d}"
        shot_file = shots_dir / f"{shot_id}.json"
        
        if not shot_file.exists():
            continue
            
        with open(shot_file, "r", encoding="utf-8") as f:
            shot_data = json.load(f)
            
        shot_out_dir = output_dir / shot_id
        shot_out_dir.mkdir(parents=True, exist_ok=True)
        
        scene = shot_data.get("scene", {})
        lighting = scene.get("lighting", {})
        camera = shot_data.get("camera", {})
        movement = camera.get("movement", {})
        chars = shot_data.get("characters", [])
        
        scene_desc = f"{scene.get('location_name', '')}。{scene.get('time_of_day', '')}。{lighting.get('description', '')}。{scene.get('atmosphere', '')}。环境：{','.join(scene.get('environmental_effects', []))}。"
        
        char_desc = ""
        char_refs = ""
        for c in chars:
            c_name = c.get('name', '')
            char_desc += f"{c_name}, {c.get('importance', 'primary')}角色。身穿{c.get('costume_id', '默认服装')}。{c.get('pose', '')}, {c.get('action', {}).get('main', '')}。情绪：{c.get('expression', {}).get('description', '')}。\n"
            char_refs += f"- {c_name}: works/穿书后我攻略了奸臣首辅/assets/characters/{c_name}/views.png（四视图，优先使用；如无则用 front.png）\n"

        scene_ref = f"- {scene.get('location_name', '未知')}: works/穿书后我攻略了奸臣首辅/assets/scenes/{scene.get('location_type', 'indoor')}/{scene.get('location_name', '未知')}/main.png\n"

        prev_end_path = f"works/穿书后我攻略了奸臣首辅/output/episode_001/shot_{shot_idx-1:03d}/end_latest.png" if shot_idx > 1 else "无，本镜为首镜"
        
        start_md = f"""【图片基本信息】
比例: {shot_data.get('global_settings', {}).get('aspect_ratio', '16:9')}
风格: {global_style}

【场景环境】
{scene_desc}

【角色设定】
{char_desc}
【角色参考图】
{char_refs}
【场景参考图】
{scene_ref}
【上一分镜尾帧】
- 路径: {prev_end_path}
- 说明: 生成首帧时必须将此图传入 generate_image 的 ImagePaths 参数，以保持场景和人物的视觉连贯性

【构图与运镜】
镜头类型: {camera.get('shot_size', '')}
镜头角度: {camera.get('angle', '')}
焦点: {camera.get('focus', {}).get('subject', '')}
景深: {camera.get('depth_of_field', '')}

【画面描述】
{shot_data.get('narrative', {}).get('purpose', '')}

【色调与光影】
主色调: {lighting.get('color', '')}
布光方式: {lighting.get('direction', '')}
对比度: 高
"""

        with open(shot_out_dir / "start.md", "w", encoding="utf-8") as f:
            f.write(start_md)
            
        with open(shot_out_dir / "middle.md", "w", encoding="utf-8") as f:
            f.write(start_md.replace('【上一分镜尾帧】', ''))

        with open(shot_out_dir / "end.md", "w", encoding="utf-8") as f:
            f.write(start_md.replace('【上一分镜尾帧】', ''))
            
        video_md = f"""【视频基本信息】
时长: {shot_data.get('duration', {}).get('seconds', 5.0)}秒
风格: {global_style}
比例: 16:9
帧率: 24fps

【场景环境】
{scene_desc}

【角色设定】
{char_desc}
【角色参考图】
{char_refs}
【运镜设计】
镜头类型: {camera.get('shot_size', '')}
镜头角度: {camera.get('angle', '')}
运镜方式: {movement.get('type', '')} - {movement.get('start_end_description', '')}

0-5秒【发生事件】:
{shot_data.get('narrative', {}).get('purpose', '')}

【色调与光影】
主色调: {lighting.get('color', '')}
布光方式: {lighting.get('direction', '')}

【台词】
说话人: {shot_data.get('dialogue', {}).get('speaker', '无')}
台词: "{shot_data.get('dialogue', {}).get('text', '无')}"
语气: {shot_data.get('dialogue', {}).get('delivery', {}).get('tone', '')}

【背景音乐】
类型: {shot_data.get('sound', {}).get('bgm', {}).get('mood', '无')}
"""
        with open(shot_out_dir / "video.md", "w", encoding="utf-8") as f:
            f.write(video_md)
            
        print(f"Generated prompts for {shot_id}")

generate_prompts()

File: generate_prompts.js
const fs = require('fs');
const path = require('path');

function generatePrompts() {
    const projectDir = "f:\\Projects\\koala\\works\\穿书后我攻略了奸臣首辅";
    const shotsDir = path.join(projectDir, "scripts", "episode_001", "shots");
    const outputDir = path.join(projectDir, "output", "episode_001");

    let globalStyle = "古风动漫风格";
    try {
        const projectData = JSON.parse(fs.readFileSync(path.join(projectDir, "project.json"), 'utf8'));
        if (projectData.style && projectData.style.globalPromptStyle) {
            globalStyle = projectData.style.globalPromptStyle;
        }
    } catch (e) { }

    for (let i = 1; i <= 4; i++) {
        const shotId = `shot_${i.toString().padStart(3, '0')}`;
        const shotFile = path.join(shotsDir, `${shotId}.json`);

        if (!fs.existsSync(shotFile)) continue;

        const shotData = JSON.parse(fs.readFileSync(shotFile, 'utf8'));
        const shotOutDir = path.join(outputDir, shotId);

        if (!fs.existsSync(shotOutDir)) {
            fs.mkdirSync(shotOutDir, { recursive: true });
        }

        const scene = shotData.scene || {};
        const lighting = scene.lighting || {};
        const camera = shotData.camera || {};
        const movement = camera.movement || {};
        const chars = shotData.characters || [];

        const envEffects = scene.environmental_effects ? scene.environmental_effects.join(',') : '';
        const sceneDesc = `${scene.location_name || ''}。${scene.time_of_day || ''}。${lighting.description || ''}。${scene.atmosphere || ''}。环境：${envEffects}。`;

        let charDesc = "";
        let charRefs = "";
        for (let c of chars) {
            const cName = c.name || '';
            const actionMain = (c.action && c.action.main) ? c.action.main : '';
            const exprDesc = (c.expression && c.expression.description) ? c.expression.description : '';
            charDesc += `${cName}, ${c.importance || 'primary'}角色。身穿${c.costume_id || '默认服装'}。${c.pose || ''}, ${actionMain}。情绪：${exprDesc}。\n`;
            charRefs += `- ${cName}: works/穿书后我攻略了奸臣首辅/assets/characters/${cName}/views.png（四视图，优先使用；如无则用 front.png）\n`;
        }

        const locName = scene.location_name || '未知';
        const locType = scene.location_type || 'indoor';
        const sceneRef = `- ${locName}: works/穿书后我攻略了奸臣首辅/assets/scenes/${locType}/${locName}/main.png\n`;

        const prevEndPath = i > 1 ? `works/穿书后我攻略了奸臣首辅/output/episode_001/shot_${(i - 1).toString().padStart(3, '0')}/end_latest.png` : "无，本镜为首镜";

        const focusSubject = (camera.focus && camera.focus.subject) ? camera.focus.subject : '';
        const narrativePurpose = (shotData.narrative && shotData.narrative.purpose) ? shotData.narrative.purpose : '';

        const startMd = `【图片基本信息】
比例: ${(shotData.global_settings && shotData.global_settings.aspect_ratio) ? shotData.global_settings.aspect_ratio : '16:9'}
风格: ${globalStyle}

【场景环境】
${sceneDesc}

【角色设定】
${charDesc}
【角色参考图】
${charRefs}
【场景参考图】
${sceneRef}
【上一分镜尾帧】
- 路径: ${prevEndPath}
- 说明: 生成首帧时必须将此图传入 generate_image 的 ImagePaths 参数，以保持场景和人物的视觉连贯性

【构图与运镜】
镜头类型: ${camera.shot_size || ''}
镜头角度: ${camera.angle || ''}
焦点: ${focusSubject}
景深: ${camera.depth_of_field || ''}

【画面描述】
${narrativePurpose}

【色调与光影】
主色调: ${lighting.color || ''}
布光方式: ${lighting.direction || ''}
对比度: 高
`;

        fs.writeFileSync(path.join(shotOutDir, "start.md"), startMd, 'utf8');
        fs.writeFileSync(path.join(shotOutDir, "middle.md"), startMd.replace('【上一分镜尾帧】\n- 路径: ' + prevEndPath + '\n- 说明: 生成首帧时必须将此图传入 generate_image 的 ImagePaths 参数，以保持场景和人物的视觉连贯性\n\n', ''), 'utf8');
        fs.writeFileSync(path.join(shotOutDir, "end.md"), startMd.replace('【上一分镜尾帧】\n- 路径: ' + prevEndPath + '\n- 说明: 生成首帧时必须将此图传入 generate_image 的 ImagePaths 参数，以保持场景和人物的视觉连贯性\n\n', ''), 'utf8');

        const speaker = (shotData.dialogue && shotData.dialogue.speaker) ? shotData.dialogue.speaker : '无';
        const text = (shotData.dialogue && shotData.dialogue.text) ? shotData.dialogue.text : '无';
        const tone = (shotData.dialogue && shotData.dialogue.delivery && shotData.dialogue.delivery.tone) ? shotData.dialogue.delivery.tone : '';
        const duration = (shotData.duration && shotData.duration.seconds) ? shotData.duration.seconds : 5.0;
        const bgmMood = (shotData.sound && shotData.sound.bgm && shotData.sound.bgm.mood) ? shotData.sound.bgm.mood : '无';
        const startEndDesc = movement.start_end_description || '';

        const videoMd = `【视频基本信息】
时长: ${duration}秒
风格: ${globalStyle}
比例: 16:9
帧率: 24fps

【场景环境】
${sceneDesc}

【角色设定】
${charDesc}
【角色参考图】
${charRefs}
【运镜设计】
镜头类型: ${camera.shot_size || ''}
镜头角度: ${camera.angle || ''}
运镜方式: ${movement.type || ''} - ${startEndDesc}

0-5秒【发生事件】:
${narrativePurpose}

【色调与光影】
主色调: ${lighting.color || ''}
布光方式: ${lighting.direction || ''}

【台词】
说话人: ${speaker}
台词: "${text}"
语气: ${tone}

【背景音乐】
类型: ${bgmMood}
`;
        fs.writeFileSync(path.join(shotOutDir, "video.md"), videoMd, 'utf8');
        console.log(`Generated prompts for ${shotId}`);
    }
}

generatePrompts();
